rules/klayout/drc/testing/run_switch_checking.py

```diff
@@ -69,61 +70,103 @@ def run_switches():
+            logging.info("## Found the expected log: " + str(out))
+            logging.info("Test case passed. {}".format(test_case))
+            logging.info(f'echo Test case passed. >> {RUN_DIRECTORY}{LOG_FILE_NAME}')
+            return True
+        elif len(out) < 1 and expected_result == PASS:
+            logging.error("Test case passed as expected but didn't generate the expected log: {}".format(test_case))
+            return False
         else:
-            logging.error("Test case FAILD check log file for more info.")
-            os.system('echo Test case FAILD check log file for more info. >> {0}{1}'.format(RUN_DIRECTORY,LOG_FILE_NAME))
+            logging.error("Test case FAILD check log file for more info. {}".format(test_case))
+            logging.error(f'echo Test case FAILD check log file for more info. >> {RUN_DIRECTORY}{LOG_FILE_NAME}')
+            return False
 
     ## read testcases from csv file
     df = pd.read_csv(csv_path)
+    pd.set_option('display.max_columns', None)
+    pd.set_option('display.max_rows', None)
+    pd.set_option("max_colwidth", None)
+    pd.set_option('display.width', 1000)
+
+    print("## All Switch checking patterns:")
+    print(df)
+
     test_cases = df["test_case_name"]
     switches = df["switches"]
     expected_results = df["expected_output"]
     expected_log = df["expected_logs"]
 
     ## init log file
-    os.system('mkdir -p {0}'.format(RUN_DIRECTORY))
-    os.system('touch {0}temp.log'.format(RUN_DIRECTORY))
-    os.system('mv -f {0}temp.log {0}{1}'.format(RUN_DIRECTORY,LOG_FILE_NAME))
+    check_call(
+        f'''
+mkdir -p {RUN_DIRECTORY}
+touch {RUN_DIRECTORY}temp.log
+mv -f {RUN_DIRECTORY}temp.log {RUN_DIRECTORY}{LOG_FILE_NAME}
+''',
+        shell=True
+    )
+
+    run_status = []
+    sw_failed = False
 
     for test_case_index in range(0,len(test_cases)):
         test_case_switches = switches[test_case_index]
 
         running_msg = "\nrunning: "+test_cases[test_case_index] + "     with switches: " + test_case_switches + "..."
         logging.info(running_msg)
-        os.system('echo "{0}" >> {1}{2}'.format(running_msg, RUN_DIRECTORY, LOG_FILE_NAME))
+        check_call(f'echo "{running_msg}" >> {RUN_DIRECTORY}{LOG_FILE_NAME}', shell=True)
+
+        case_res = run_test_case(test_cases[test_case_index], test_case_switches, expected_results[test_case_index],expected_log[test_case_index])
+        run_status.append(case_res)
 
-        run_test_case(test_cases[test_case_index], test_case_switches, expected_results[test_case_index],expected_log[test_case_index])
+        if not case_res:
+            sw_failed = True
 
+    df["run_status"] = run_status
+    
+    if sw_failed:
+        logging.error("## One of the test cases failed. Exit with failure:")
+        print(df)
+        exit(1)
+    else:
+        logging.info("## All test cases passed.")
+        logging.info("## All Switch checking patterns:")
+        print(df)
```

Comment:
`logging`?

Reply:
@proppy what's the difference?

Reply:
`logging` gives you more control over the leveling of the log and allow you to optionally direct the output to a file, see https://docs.python.org/3/library/logging.html

Reply:
@proppy I got the point. Yes, that's why we use logging in the first place. The same comment as yours. But based on my experience printing Pandas DataFrame is nicer in print rather than logging. I could change it. But it won't be nice in the command line.